receiver.go

```diff
@@ -108,39 +52,113 @@ func (r *Receiver) Receive(ctx context.Context) (*Message, error) {
 		}
 	}
 
+	// for ModeFirst, auto-accept the message
+	autoAccept := func(ctx context.Context, msg *Message) (*Message, error) {
+		if receiverSettleModeValue(r.link.ReceiverSettleMode) == ModeSecond {
+			return msg, nil
+		}
+		if err := r.AcceptMessage(ctx, msg); err != nil {
+			return nil, err
+		}
+		return msg, nil
+	}
+
 	// non-blocking receive to ensure buffered messages are
 	// delivered regardless of whether the link has been closed.
 	select {
 	case msg := <-r.link.Messages:
-		// we remove the message from unsettled map as soon as it's popped off the channel
-		// This makes the unsettled count the same as messages buffer count
-		// and keeps the behavior the same as before the unsettled messages tracking was introduced
-		defer r.link.DeleteUnsettled(&msg)
 		debug(3, "Receive() non blocking %d", msg.deliveryID)
-		msg.receiver = r
-		return &msg, nil
+		msg.link = r.link
+		return autoAccept(ctx, &msg)
```

Comment:
This gives the impression that we always autoaccept. I had to look at the implementation of autoaccept to see that it only accept if not mode second. Can probably use a better func name.

Reply:
Good point, renamed.